### PR TITLE
Flag to switch between static and dynamic length of differentiable parameters

### LINKED ISSUE #26: Support optimisers storing interim results
Some optimisers such as the ADAM optimiser have a memory to store recent results as part of their optimisation concept to improve on other methods such as Gradient Descent. However, currently our project does change the number of parameters relevant for optimisation during the training process. This in turn either creates false assumptions with optimisers having a memory as arrays don't map on each other or in the extreme case lead to an exception during training (e.g. out of bounds errors as length of arrays might mismatch).

To cope with this, we might either use `reset` functionality on the optimisers. Based on the assumption that our count of parameters only changes whenever the gradients are close to `0`, we might assume that interim results of optimisers are meaningless. However, this must be checked with the different implementations of optimisers.

Further, we might introduce the capability to *freeze* values to `0` when we initially wanted to remove them, so that the changing action from a given optimiser does not take place and is plainly ignored from our side. However, for this use case it might happen, that interim values represent another reality as the optimiser might assume that its requested change took place. Here we again have to check implementation of optimisers.

### PR DESCRIPTION
**Pull request by eileen-kuehn**
This PR implements the feature to switch between dynamic and static length of differentiable parameters and therefore allows to use optimisers storing intermediate values again.
For ADAM the flag is set to `False` to enable a static length of parameters to be differentiated. This has been implemented as a default as small experiments showed that too many information are lost when the optimiser is constantly reseted for each branch. This resulted in a worse convergence time.

Closes #26.

### CODE REVIEW

File: maskit/masks.py [hunk 1]
```diff
@@ -172,6 +172,11 @@ class MaskedCircuit(object):
     """
     A MaskedCircuit supports masking of different components including wires, layers,
     and parameters.
+
+    :param dynamic_parameters: The array of differentiable parameters changes over
+        time depending on the given masks. In case `True` is supplied, the least
+        amount of parameters is returned, otherwise the length is fixed and independent
+        of masks
```
Comment (maxfischer2781) [inline maskit/masks.py:179]:
```suggestion

    :param dynamic_parameters: Whether the array of differentiable parameters may change size/order
    
    Masking naturally removes active parameters from a circuit.
    However, some optimisers expect the array of parameters to
    remain stable across iteration steps;
    use ``dynamic_parameters=False`` to force the mask to always
    yield the full set of parameters in such cases.
    The mask will still prevent modification of inactive parameters.
```

**Review by maxfischer2781 — CHANGES_REQUESTED** [review]
Just got a suggestion on the description of the feature. The current one wasn't clear to me and does not seem actionable.

**Review by maxfischer2781 — APPROVED** [review]
Oops, sorry for putting the suggestion to the wrong location! Good to go now. 👍
